src/storage/UpdateVertexProcessor.cpp

```diff
@@ -0,0 +1,344 @@
+                auto v = boost::get<double>(expValue);
+                tagUpdaters_[tagId]->updater->setDouble(prop, v);
+                break;
+            }
+            case VAR_BOOL: {
+                auto v = boost::get<bool>(expValue);
+                tagUpdaters_[tagId]->updater->setBool(prop, v);
+                break;
+            }
+            case VAR_STR: {
+                auto v = boost::get<std::string>(expValue);
+                tagUpdaters_[tagId]->updater->setString(prop, v);
+                break;
+             }
+            default: {
+                LOG(FATAL) << "Unknown VariantType: " << expValue.which();
+                return std::string("");
+            }
+        }
+    }
+
+    std::vector<kvstore::KV> data;
+    for (const auto& u : tagUpdaters_) {
+        data.emplace_back(std::move(u.second->key),
+                          std::move(u.second->updater->encode()));
+    }
+    auto log = kvstore::encodeMultiValues(kvstore::OP_MULTI_PUT, data);
+    return log;
+}
+
+
+cpp2::ErrorCode UpdateVertexProcessor::checkAndBuildContexts(
+        const cpp2::UpdateVertexRequest& req) {
+    if (this->expCtx_ == nullptr) {
+        this->expCtx_ = std::make_unique<ExpressionContext>();
+    }
+
+    // return columns
+    for (auto& col : req.get_return_columns()) {
+        StatusOr<std::unique_ptr<Expression>> colExpRet = Expression::decode(col);
+        if (!colExpRet.ok()) {
+            return cpp2::ErrorCode::E_INVALID_UPDATER;
+        }
+        auto colExp = std::move(colExpRet).value();
+        colExp->setContext(this->expCtx_.get());
+        auto status = colExp->prepare();
+        if (!status.ok() || !this->checkExp(colExp.get())) {
+            return cpp2::ErrorCode::E_INVALID_UPDATER;
+        }
+        returnColumnsExp_.emplace_back(std::move(colExp));
+    }
+
+    // filter(where/when)
+    const auto& filterStr = req.get_filter();
+    if (!filterStr.empty()) {
+        StatusOr<std::unique_ptr<Expression>> expRet = Expression::decode(filterStr);
+        if (!expRet.ok()) {
+            VLOG(1) << "Can't decode the filter " << filterStr;
+            return cpp2::ErrorCode::E_INVALID_FILTER;
+        }
+        this->exp_ = std::move(expRet).value();
+        this->exp_->setContext(this->expCtx_.get());
+        auto status = this->exp_->prepare();
+        if (!status.ok() || !this->checkExp(this->exp_.get())) {
+            return cpp2::ErrorCode::E_INVALID_FILTER;
+        }
+    }
+
+    // build context of the update items
+    for (auto& item : req.get_update_items()) {
+        std::string name = item.get_name();
+        std::string prop = item.get_prop();
+        auto* sourcePropExp = new SourcePropertyExpression(&name, &prop);
+        sourcePropExp->setContext(this->expCtx_.get());
+        auto status = sourcePropExp->prepare();
+        if (!status.ok() || !this->checkExp(sourcePropExp)) {
+            return cpp2::ErrorCode::E_INVALID_UPDATER;
```

Comment:
before return, maybe need delete sourcePropExp, else memory leak.

Reply:
We can't delete it, because all members of Expression are smart ptr.
FYI.  https://github.com/vesoft-inc/nebula/blob/master/src/storage/QueryBaseProcessor.inl#L249